### ISSUE TITLE
Rails 4.1 Stepping Through Associations

### ISSUE THREAD

**Issue by ngottlieb** [pat/thinking-sphinx#978]
Hey Pat,

I'm trying to upgrade an app to Rails 4.1 (formerly on 3.2, but it worked on 4.0 as a step up to 4.1). Thinking-sphinx `3.1.4`. Here's the issue I'm running into: formerly, on Rails 3.2, and 4.0, this piece of an index definition:

    has [locations.id, organization.locations.id], as: :all_loc_ids, multi: true

generated this piece of the SQL query in the Sphinx configuration:

    GROUP_CONCAT(DISTINCT CONCAT_WS(',', CAST(`locations`.`id` AS char), CAST(`locations_organizations`.`id` AS char)) SEPARATOR ',') AS `all_loc_ids`,

so far so good. I upgrade to Rails 4.1 and instead, I get this, which throws no errors but doesn't really function how I need it to:

    GROUP_CONCAT(DISTINCT CONCAT_WS(',', CAST(`locations`.`id` AS char), CAST(`locations`.`id` AS char)) SEPARATOR ',') AS `all_loc_ids`,

At first I thought it was a syntax thing, maybe that way of representing things isn't actually supported, so I tried 

    has organization.locations.id, as: :all_loc_ids, multi: true

but unfortunately, that yielded:

    GROUP_CONCAT(DISTINCT `locations`.`id` SEPARATOR ',') AS `all_loc_ids`

I've dug around a bit and can't figure this out. It's presumably a difference in ActiveRecord that's causing this. My theory is that this problem only crops up when you have an attribute on the associated model (or another association) with the same name as an attribute (or association) on the base model. This issue might belong on the `joiner` gem but I can't really give context specific enough without the `thinking-sphinx` index context.

Any ideas?

Thanks.

**Comment by pat** [comment]
Hi Nick

This is certainly surprising! Have you tried running `bundle update joiner` to see if that helps at all? Also, can you talk me through the associations in play here? i.e. which model is the index tied to, I presume it has_many locations and belongs_to organization, which in turn has_many locations, though they are a different model rather than Location?

**Comment by ngottlieb** [comment]
Yeah, I've tried `bundle update joiner`, no change. We're on `0.3.4`. 

The index is on the Content model. Content `has_and_belongs_to_many :locations` and `belongs_to :organization`. Organization *also* `has_and_belongs_to_many :locations`.

I'll try to generate an app from scratch and see if I can duplicate the issue outside our app.

**Comment by ngottlieb** [comment]
Interesting. I just generated an app with what seems to be the same configuration with TS 3.1.4, Rails 4.1 and lo and behold -- it worked fine:

    GROUP_CONCAT(DISTINCT CONCAT_WS(',', CAST(`locations`.`id` AS char), CAST(`locations_organizations`.`id` AS char))  SEPARATOR ',') AS `all_loc_ids`

so I guess there's something else in our configuration that is causing this to break with Rails 4.1 (it definitely worked on 4.0 and nothing else has changed). I'll dig into it and update if I can figure it out...

**Comment by ngottlieb** [comment]
Alright! I isolated the issue. It's in the index file. Basically, this index works fine:

    ThinkingSphinx::Index.define :content, with: :active_record do
      has [locations.id, organization.locations.id], as: :all_loc_ids, multi: true
    end

and generates the correct `all_loc_ids` query snippet (seen in my last comment).

*This* index file does not work:

    ThinkingSphinx::Index.define :content, with: :active_record do
      has organization.id, as: :org_id
      has [locations.id, organization.locations.id], as: :all_loc_ids, multi: true
    end

Of note, this problem only started with Rails 4.1. Here are the full queries generated by both indices:

Working index (the first one):

```
SELECT SQL_NO_CACHE `contents`.`id` * 2 + 0 AS `id`, `contents`.`id` AS `sphinx_internal_id`, 'Content' AS `sphinx_internal_class`, 0 AS `sphinx_deleted`, GROUP_CONCAT(DISTINCT CONCAT_WS(',', CAST(`locations`.`id` AS char), CAST(`locations_organizations`.`id` AS char)) SEPARATOR ',') AS `all_loc_ids` FROM `contents` LEFT OUTER JOIN `contents_locations` ON `contents_locations`.`content_id` = `contents`.`id` LEFT OUTER JOIN `locations` ON `locations`.`id` = `contents_locations`.`location_id` LEFT OUTER JOIN `organizations` ON `organizations`.`id` = `contents`.`organization_id` LEFT OUTER JOIN `locations_organizations` `organizations_locations_organizations_join` ON `organizations_locations_organizations_join`.`organization_id` = `organizations`.`id` LEFT OUTER JOIN `locations` `locations_organizations` ON `locations_organizations`.`id` = `organizations_locations_organizations_join`.`location_id` WHERE (`contents`.`id` BETWEEN $start AND $end) GROUP BY `contents`.`id`, `contents`.`id`  ORDER BY NULL
```

Incorrect index (second): 

```
SELECT SQL_NO_CACHE `contents`.`id` * 2 + 0 AS `id`, `contents`.`id` AS `sphinx_internal_id`, 'Content' AS `sphinx_internal_class`, 0 AS `sphinx_deleted`, `organizations`.`id` AS `org_id`, GROUP_CONCAT(DISTINCT CONCAT_WS(',', CAST(`locations`.`id` AS char), CAST(`locations`.`id` AS char)) SEPARATOR ',') AS `all_loc_ids` FROM `contents` LEFT OUTER JOIN `organizations` ON `organizations`.`id` = `contents`.`organization_id` LEFT OUTER JOIN `locations_organizations` ON `locations_organizations`.`organization_id` = `organizations`.`id` LEFT OUTER JOIN `locations` ON `locations`.`id` = `locations_organizations`.`location_id` LEFT OUTER JOIN `contents_locations` ON `contents_locations`.`content_id` = `contents`.`id` LEFT OUTER JOIN `locations` `locations_contents` ON `locations_contents`.`id` = `contents_locations`.`location_id` WHERE (`contents`.`id` BETWEEN $start AND $end) GROUP BY `contents`.`id`, `contents`.`id`, `organizations`.`id`  ORDER BY NULL
  sql_query_range = SELECT IFNULL(MIN(`contents`.`id`), 1), IFNULL(MAX(`contents`.`id`), 1) FROM `contents`
```

Any ideas? It does seem to be a TS / joiner issue after all.

**Comment by ngottlieb** [comment]
I was looking at the `joiner` repo and found this vaguely similar sounding issue -- https://github.com/pat/joiner/issues/3 -- so I tried flipping the order of those two calls and it started working. So:

```
ThinkingSphinx::Index.define :content, with: :active_record do
  has [locations.id, organization.locations.id], as: :all_loc_ids, multi: true
  has organization.id, as: :org_id
end
```

correctly generates the all_loc_ids query while the previous arrangement does not. 

I don't think this bug is actually the same (especially since you said you fixed that one), but the flipping the order solved it.

**Comment by pat** [comment]
Great to hear you found a solution! Now I guess the challenge is to figure out a fix… ActiveRecord is such fun to deal with :|

**Comment by ngottlieb** [comment]
Hahah, I'd love to help, but without a bit of direction I'm definitely going to be in over my head trying to parse what joiner, ActiveRecord, etc., are all doing.

**Comment by pat** [comment]
I found the bug and fixed this!

When you're constructing a SQL query with ActiveRecord and adding joins one by one, you can then ask for table aliases. However, if you then add more joins to the same tables in different contexts (as you're doing with the multiple references to your `locations` table) the aliases change (and it seems the most recent references get the actual table name, where as the joins added earlier get prefixed versions). Thinking Sphinx was lazily adding joins and finding aliases, and thus was dealing with a moving target.

So, Thinking Sphinx _now_ adds all joins at the very beginning of SQL generation, thus ensuring that the aliases don't change. This is part of the just-published [3.2.0](https://github.com/pat/thinking-sphinx/releases/tag/v3.2.0) release :)

**Comment by ngottlieb** [comment]
Awesome, nice work! Thanks!
